File: php/testdriver.php
<?php

// Test Driver
//
//

function readtestdata($filename) {
    $state = '';
    $info = array();

    $fd = fopen($filename, 'r');
    while (! feof($fd)) {
       $line = trim(fgets($fd));
       if ($line == '--TEST--' || $line == '--INPUT--' || $line == '--EXPECTED--') {
           $state = $line;
           $info[$state] = '';
       } else if ($state != '') {
           $info[$state] .= $line . "\n";
       }
    }
    fclose($fd);

    return array(
        trim($info['--TEST--']),
        trim($info['--INPUT--']),
        trim($info['--EXPECTED--'])
    );
}

function fileproviders($pattern) {
    $args = array();
    foreach (glob($pattern) as $a) {
        array_push($args, array($a));
    }
    return $args;
}


class LibinjectionTestTokens extends PHPUnit_Framework_TestCase
{
    /**
     * @dataProvider filesTokens
     */
    public function testTokens($testfile) {
        return $this->_run($testfile, 'tokens', FLAG_QUOTE_NONE | FLAG_SQL_ANSI);
    }
    public function filesTokens() {
        return fileproviders('../tests/test-tokens-*.txt');
    }

    /**
     * @dataProvider filesFolding
     */
    public function testFolding($testfile) {
         return $this->_run($testfile, 'folding', FLAG_QUOTE_NONE | FLAG_SQL_ANSI);
    }
    public function filesFolding() {
        return fileproviders('../tests/test-folding-*.txt');
    }

    /**
     * @dataProvider filesFingerprints
     */
    public function testFingerprints($testfile) {
        return $this->_run($testfile, 'fingerprints', 0);
    }
    public function filesFingerprints() {
        return fileproviders('../tests/test-sqli-*.txt');
    }

    public function _run($testname, $flag, $sqli_flags) {
        $data =  readtestdata($testname);
        $sqlistate = new_libinjection_sqli_state();
        libinjection_sqli_init($sqlistate, $data[1], $sqli_flags);
        $actual = '';
        if ($flag == 'tokens') {
            while (libinjection_sqli_tokenize($sqlistate)) {
                $actual .= $this->print_token(libinjection_sqli_state_current_get($sqlistate)) . "\n";
            }
        } else if ($flag == 'folding') {
            $fingerprint = libinjection_sqli_fingerprint($sqlistate, $sqli_flags);
            for ($i  = 0; $i < strlen($fingerprint); $i++) {
                $r = libinjection_sqli_state_tokenvec_geti($sqlistate, $i);
                $actual .= $this->print_token($r) . "\n";
            }
        } else if ($flag == 'fingerprints') {
            $ok = libinjection_is_sqli($sqlistate);
            if ($ok == 1) {
                $actual = libinjection_sqli_state_fingerprint_get($sqlistate);
            }
        } else {
           $this->assert(False);
        }
        $actual = trim($actual);
        $this->assertEquals($actual, $data[2]);
    }

    public function print_token($tok) {
        $tt = libinjection_sqli_token_type_get($tok);
        $out = '';
        $out .= $tt;
        $out .= ' ';
        if ($tt == 's') {
            $out .= $this->print_token_string($tok);
        } else if ($tt == 'v') {
            $vc = libinjection_sqli_token_count_get($tok);
            if ($vc == 1) {
                $out .= '@';
            } else if ($vc == 2) {
                $out .= '@@';
            }
            $out .= $this->print_token_string($tok);
        } else {
            $out .= libinjection_sqli_token_val_get($tok);
        }
        return trim($out);
    }

    public function print_token_string($tok) {
       $out = '';
       $quote = libinjection_sqli_token_str_open_get($tok);
       if ($quote != "\0") {
           $out .= $quote;
       }
       $out .= libinjection_sqli_token_val_get($tok);
       $quote = libinjection_sqli_token_str_close_get($tok);
       if ($quote != "\0") {
           $out .= $quote;
       }
       return $out;
    }
}

File: src/test_driver.py
#!/usr/bin/env python
"""
Test driver
Runs off plain text files, similar to how PHP's test harness works
"""
import subprocess
import os
import glob

def toascii(data):
    """
    Converts a utf-8 string to ascii.
    Needed since nosetests xunit is not UTF-8 safe
    https://github.com/nose-devs/nose/issues/649
    https://github.com/nose-devs/nose/issues/692
    """
    udata = data.decode('utf-8')
    return udata.encode('ascii', 'xmlcharrefreplace')

def run(args):
    """
    Runs a command and returns stdout output, throws exception on failed run.
    Uses python's 2.7 check_output if available otherwise uses something else
    """

    try:
        getattr(subprocess, 'check_output')
        has_check_output = True
    except AttributeError:
        has_check_output = False

    if has_check_output:
        # 2.7
        # turn off warning on 2.6 systems
        # pylint: disable=E1101
        return subprocess.check_output(args)
    else:
        # not 2.7
        aprocess = subprocess.Popen(args,
                                    stdout=subprocess.PIPE,
                                    stderr=subprocess.PIPE)
        (stdoutdata, stderrdata) = aprocess.communicate()
        if aprocess.returncode != 0:
            raise Exception("Test died!: " + stderrdata)
        return stdoutdata

def readtestdata(filename):
    """
    Read a test file and split into components
    """

    state = None
    info = {
        '--TEST--': '',
        '--INPUT--': '',
        '--EXPECTED--': ''
        }

    for line in open(filename, 'r'):
        line = line.rstrip()
        if line in ('--TEST--', '--INPUT--', '--EXPECTED--'):
            state = line
        elif state:
            info[state] += line + '\n'

    # remove last newline from input
    info['--INPUT--'] = info['--INPUT--'][0:-1]

    return (
        info['--TEST--'],
        info['--INPUT--'].strip(),
        info['--EXPECTED--'].strip()
    )

def runtest(testname, flags=None):
    """
    runs a test, optionally with valgrind
    """
    data =  readtestdata(os.path.join('../tests', testname))

    if os.environ.get('VALGRIND', None):
        args = [
            'valgrind',
            '--gen-suppressions=no',
            '--read-var-info=yes',
            '--error-exitcode=1',
            '--track-origins=yes',
            ]
    else:
        args = []

    args.append(os.getenv('PARSER_CMD', './sqli'))

    if flags:
        for flag in flags:
            args.append(flag)

    args.append(data[1])
    actual = run(args)

    actual = actual.strip()

    if actual != data[2]:
        print "INPUT: \n" + toascii(data[1])
        print
        print "EXPECTED: \n" + toascii(data[2])
        print
        print "GOT: \n" + toascii(actual)
        assert actual == data[2]

def run_tokens(testname):
    """ run a single generic SQL token parsing test """
    runtest(testname, [ '-q0', '-ca'])

def run_tokens_mysql(testname):
    """ run a single MySQL token parsing test """
    runtest(testname, [ '-q0', '-cm'])

def run_folding(testname):
    """ run a single folding test """
    runtest(testname, ['-f', '-q0', '-ca'])

def run_sqli(testname):
    """ run a single sqli detection test """
    runtest(testname, ['-d', ])

def test_tokens():
    """ wrapper for generic SQL token parsing """
    for testname in sorted(glob.glob('../tests/test-tokens-*.txt')):
        testname = os.path.basename(testname)
        yield run_tokens, testname

def test_tokens_mysql():
    """ testing for mysql specific parsing """
    for testname in sorted(glob.glob('../tests/test-tokens_mysql-*.txt')):
        testname = os.path.basename(testname)
        yield run_tokens_mysql, testname

def test_folding():
    """ wrapper for testing folding """
    for testname in sorted(glob.glob('../tests/test-folding-*.txt')):
        testname = os.path.basename(testname)
        yield run_folding, testname

def test_sqli():
    """ wrapper for testing high level is_sqli """
    for testname in sorted(glob.glob('../tests/test-sqli-*.txt')):
        testname = os.path.basename(testname)
        yield run_sqli, testname

if __name__ == '__main__':
    import sys
    sys.stderr.write("run using nosetests\n")
    sys.exit(1)
